File: src/test/java/io/github/bbwdadfg/atwr/AtwrToolsTest.java
package io.github.bbwdadfg.atwr;

import static org.junit.jupiter.api.Assertions.assertEquals;

import java.util.List;
import org.junit.jupiter.api.Test;

class AtwrToolsTest {
    @Test
    void scanReportsCodepointsAndPositions() {
        List<AtwrTools.Finding> findings = AtwrTools.scanInvisibleCharacters("A\u200b\u202fB");
        assertEquals(2, findings.size());
        assertEquals("U+200B", findings.get(0).codePoint);
        assertEquals(1, findings.get(0).index);
        assertEquals("U+202F", findings.get(1).codePoint);
    }

    @Test
    void removePreservesVisibleText() {
        assertEquals("A B", AtwrTools.removeInvisibleCharacters("A\u200b B\ufeff"));
    }

    @Test
    void tidyMarkdownResidue() {
        assertEquals("Heading\nItem\n[link]", AtwrTools.stripMarkdownPasteResidue("## Heading\n- **Item**\n\\[link\\]"));
    }
}

File: atwrtools_test.go
package atwrtools

import "testing"

func TestScanInvisibleCharacters(t *testing.T) {
	findings := ScanInvisibleCharacters("A\u200b\u202fB")
	if len(findings) != 2 || findings[0].CodePoint != "U+200B" || findings[0].Index != 1 || findings[1].CodePoint != "U+202F" {
		t.Fatalf("unexpected findings: %#v", findings)
	}
}

func TestRemoveInvisibleCharacters(t *testing.T) {
	if got := RemoveInvisibleCharacters("A\u200b B\ufeff"); got != "A B" {
		t.Fatalf("got %q", got)
	}
}

func TestStripMarkdownPasteResidue(t *testing.T) {
	got := StripMarkdownPasteResidue("## Heading\n- **Item**\n\\[link\\]")
	if got != "Heading\nItem\n[link]" {
		t.Fatalf("got %q", got)
	}
}
